lib/src/memory/cache.dart

```diff
@@ -0,0 +1,487 @@
+// Copyright (C) 2024-2025 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// cache.dart
+// Set-associative cache.
+//
+// 2025 September 10
+// Author: Desmond Kirkpatrick <desmond.a.kirkpatrick@intel.com>
+
+import 'package:meta/meta.dart';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// An interface to a cache memory that supplies enable [en], address [addr],
+/// [valid] for indicating a hit, and [data].
+///
+/// Can be used for either read or write direction by grouping signals using
+/// [DataPortGroup].
+class ValidDataPortInterface extends DataPortInterface {
+  /// The "valid" bit for a response when the data is valid.
+  Logic get valid => port('valid');
+
+  /// Constructs a new interface of specified [dataWidth] and [addrWidth] for
+  /// interacting with a `Cache` in either the read or write direction.
+  ValidDataPortInterface(super.dataWidth, super.addrWidth) : super() {
+    setPorts([
+      Logic.port('valid'),
+    ], [
+      DataPortGroup.data
+    ]);
+  }
+
+  /// Makes a copy of this [ValidDataPortInterface] with matching configuration.
+  @override
+  ValidDataPortInterface clone() =>
+      ValidDataPortInterface(dataWidth, addrWidth);
+}
+
+/// A module [Cache] implementing a configurable set-associative cache. Three
+/// primary operations:
+/// - Reading from a cache can result in a hit or miss. The only state change is
+///   that on a hit, the [replacement] policy is updated.  This is similar to a
+///   memory, except that a valid bit is returned with the read data.
+/// - Writing to a cache with a valid bit set results in a write into the cache
+///   memory, potentially allocating a line in a new way if the data was not
+///   present. Externally, this just looks like a memory write.
+/// - Writing to a cache without the valid bit set results in an invalidate of
+///   the matching line if present.
+abstract class Cache extends Module {
```

Comment:
how can you tell when something is getting evicted, what got evicted?

Reply:
yeah somehow it seems important to know that information right?  maybe we can omit the feature in this first version, but its also worth thinking about whether it's an easy API to add later (without breaking backwards compatibility)

Reply:
I thought about this more.  I think we can add an eviction interface (same as Read) which is #reads + #writes wide.
Upon a miss, we can fill that addr/data with the evicted data.  Upon a write/miss, we can setup the write to that location to occur concurrently with the eviction read.  for 1R/1W cache, this would mean 2R/1W memories to support.